nupic/research/frameworks/pytorch/distributed_sampler.py

```diff
@@ -0,0 +1,108 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import torch
+import torch.distributed as dist
+from torch.utils.data import Sampler
+
+
+class UnpaddedDistributedSampler(Sampler):
+    """
+    A fork of the pytorch DistributedSampler that doesn't repeat data, instead
+    allowing the number of batches per process to be off-by-one from each other.
+    This makes this sampler usable for validation (it's deterministic and
+    doesn't require shuffling). Diffs to the code are annotated with "MODIFY"
+    comments.
+
+    Sampler that restricts data loading to a subset of the dataset.
+    It is especially useful in conjunction with
+    :class:`torch.nn.parallel.DistributedDataParallel`. In such case, each
+    process can pass a DistributedSampler instance as a DataLoader sampler,
+    and load a subset of the original dataset that is exclusive to it.
+    .. note::
+        Dataset is assumed to be of constant size.
+    Arguments:
+        dataset: Dataset used for sampling.
+        num_replicas (optional): Number of processes participating in
+            distributed training.
+        rank (optional): Rank of the current process within num_replicas.
+        shuffle (optional): If true (default), sampler will shuffle the indices
+    .. warning::
+        In distributed mode, calling the ``set_epoch`` method is needed to
+        make shuffling work; each process will use the same random seed
+        otherwise.
+    Example::
+        >>> sampler = DistributedSampler(dataset) if is_distributed else None
```

Comment:
Did you mean `UnpaddedDistributedSampler`?

Reply:
Good point. I was preserving the original pytorch docstring (this is copy-pasted), but you're right that it's better to change these.